### ISSUE TITLE
Tests

### ISSUE THREAD

**Issue by vulnguard** [ProAlgos/ProAlgos-Cpp#30]
I'm interested in writing test code for these algorithms, will as much code coverage as possible. Is that something you would like to have in this project?

**Comment by faheel** [comment]
Yes @Number1n00b, I too was thinking about adding unit tests and using TDD approach (and also measuring and increasing code coverage), but I do not have much idea as to how to implement these. So I decided to look up books on this topic and found [one](https://pragprog.com/book/lotdd/modern-c-programming-with-test-driven-development), which I'm going through currently. Only after reading some topics would I be able to decide whether, and how, tests be added and made use of effectively, without much changes to already existing code.

If you have any knowledge or experience regarding testing C++ code, then please do share and give your opinion on how should we work on this issue.

**Comment by vulnguard** [comment]
The only experience i have with code testing is in Java using JUnit, a framework designed to allow you to conduct multiple unit tests with relative ease.

I'm interested in applying my knowledge of JUnit testing to a CPP framework, although I have not researched it much yet. I've heard good things about Visual Studio for code testing as it has integrated code coverage. 

I think the best thing I can do is to start writing some test code for some of the algorithms, using Visual Studio, see how I go, and then get back to you with my progress.
I may also try to use the inbuilt testing framework of CLion (By NetBeans) if I get the chance.

**Comment by faheel** [comment]
Keep in mind that the testing framework should be cross-platform, lightweight and easy to install on all major platforms (Linux, Windows, macOS).

Using Visual Studio would make the project platform-dependent since Visual Studio is only available for Windows.

I suggest you take a look at [Google Test](https://github.com/google/googletest) and [Boost Test Library](http://www.boost.org/doc/libs/1_46_0/libs/test/doc/html/index.html).

**Comment by vulnguard** [comment]
Okay, thanks. I'll have a look into those.

**Comment by connorkuehl** [comment]
I've been using [Catch](https://github.com/philsquared/Catch), a header-only testing framework and it's been a breeze to get started up, writing and running tests. Might be worth looking into.

**Comment by faheel** [comment]
Catch seems to be perfect for the job! Thanks for the suggestion @connorkuehl! :+1: 

I'm not sure how well it integrates with systems like [Travis CI](https://travis-ci.org/) or [Codecov](https://codecov.io/), but that's something for later. Right now, for local testing, I'll add the single-header version of Catch to the repo, along with simple unit tests for either binary search or Fibonacci.

**Comment by faheel** [comment]
@Number1n00b I have added the Catch header and unit tests for Fibonacci. To get an idea what changes / additions were done for it, you can take a look at [this commit](../commit/14838e27b7192ec0df60b3c5f3baaf2adcbfeae3).

**Comment by vulnguard** [comment]
@faheel @connorkuehl Catch looks great! Thanks for the suggestion. 

I've been quite bust with university at the moment but I'm excited to write some test cases with catch once I have some more free time.

**Comment by faheel** [comment]
I'm closing this and creating a new issue to keep track of tests.
